### PR TITLE
Add e2e test stack

### PR DESCRIPTION
**Pull request by outscale-vbr**

### CODE REVIEW

File: .github/workflows/kind.yaml [hunk 1]
```diff
@@ -0,0 +1,96 @@
+name: kind
+
+on:
+  push:
+    tags:
+      - "v*"
+jobs:
+  kind:
+    runs-on: [self-hosted, linux, kind]
```
Comment (outscale-mdr) [inline .github/workflows/kind.yaml:9]: Are you sure about the labels ?

File: .github/workflows/conformance.yaml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      run:  KUBECONFIG=$GITHUB_WORKSPACE/rke-cluster-for-cluster-api/rke/kube_config_cluster.yml make e2econformance
+      shell: bash
+      id: e2etest
+      env:
+        IMG: ${{ env.NODEPORT_IP }}:${{ env.NODEPORT_PORT }}/cluster-api-outscale-controller:${{ github.sha }}
+        OSC_ACCESS_KEY: ${{secrets.OSC_ACCESS_KEY}}
+        OSC_SECRET_KEY: ${{secrets.OSC_SECRET_KEY}}
+    - name: Get cluster api outscale logs
+      run: make logs-capo
+      if: ${{ steps.e2etest.outcome == 'success' }}
```
Comment (outscale-mdr) [inline .github/workflows/conformance.yaml:121]: You only want the logs when it is a success ?

File: api/v1beta1/oscmachine_webhook.go [hunk 3]
```diff
@@ -64,15 +64,6 @@ func (m *OscMachine) ValidateUpdate(oldRaw runtime.Object) error {
...
-	oscMachineLog.Info("validate update old imageId", "old imageId", old.Spec.Node.Vm.ImageId)
-	oscMachineLog.Info("validate update imageId", "imageId", m.Spec.Node.Vm.ImageId)
-	if !reflect.DeepEqual(m.Spec.Node.Vm.ImageId, old.Spec.Node.Vm.ImageId) {
-		allErrs = append(allErrs,
-			field.Invalid(field.NewPath("spec", "imageId"),
-				m.Spec.Node.Vm.ImageId, "field is immutable"),
-		)
-	}
-
 	oscMachineLog.Info("validate update old vmType", "old vmType", old.Spec.Node.Vm.VmType)
```
Comment (outscale-mdr) [inline api/v1beta1/oscmachine_webhook.go:67]: Explain why you remove this part of the code

File: api/v1beta1/types.go [hunk 4]
```diff
@@ -313,6 +315,12 @@ type OscVm struct {
 	ClusterName        string                    `json:"clusterName,omitempty"`
 }
 
+type OscRootDisk struct {
+	RootDiskIops int32  `json:"rootDiskIops,omitempty"`
+	RootDiskSize int32  `json:"rootDiskSize, omitempty"`
+	RootDiskType string `json:"rootDiskType, omitempty"`
```
Comment (outscale-mdr) [inline api/v1beta1/types.go:321]: Remove the space between ',' and 'omitempty'

File: api/v1beta1/types.go [hunk 5]
```diff
@@ -761,6 +785,10 @@ func (network *OscNetwork) SetSecurityGroupDefaultValue() {
 		var securityGroupRuleApiKwName string = DefaultSecurityGroupRuleApiKwName
 		var securityGroupRuleApiKcpName string = DefaultSecurityGroupRuleApiKcpName
 		var securityGroupRuleEtcdName string = DefaultSecurityGroupRuleEtcdName
+		var securityGroupRuleKwBgpKwName string = DefaultSecurityGroupRuleKwBgpKwName
+                var securityGroupRuleKwBgpKcpName string = DefaultSecurityGroupRuleKwBgpKcpName
+                var securityGroupRuleKcpBgpKwName string = DefaultSecurityGroupRuleKcpBgpKwName
+		var securityGroupRuleKcpBgpKcpName string = DefaultSecurityGroupRuleKcpBgpKcpName
```
Comment (outscale-mdr) [inline api/v1beta1/types.go:791]: Format

File: cloud/services/compute/vm.go [hunk 6]
```diff
@@ -162,3 +184,20 @@ func (s *Service) GetVmState(vmId string) (string, error) {
...
+	err := tag.AddTag("OscK8SNodeName", hostname, resourceIds, oscApiClient, oscAuthClient)
+	if err != nil {
+		return fmt.Errorf("%w failed to add OscK8sNodeName tag", err)
+	}
+	err = tag.AddTag("OscK8sClusterID/"+netName, "owned", resourceIds, oscApiClient, oscAuthClient)
+	if err != nil {
+		return fmt.Errorf("%w failed to add OscK8sClusterId tag", err)
+	}
+	return nil
+}
```
Comment (outscale-mdr) [inline cloud/services/compute/vm.go:203]: This should be in its own commit

File: cloud/services/net/net.go [hunk 7]
```diff
@@ -40,6 +40,11 @@ func (s *Service) CreateNet(spec *infrastructurev1beta1.OscNet, netName string)
 		fmt.Printf("Error with http result %s", httpRes.Status)
 		return nil, err
 	}
+	err = tag.AddTag("OscK8sClusterID/"+netName, "owned", resourceIds, oscApiClient, oscAuthClient)
+	if err != nil {
+		fmt.Printf("Error with http result %s", httpRes.Status)
+		return nil, err
+	}
```
Comment (outscale-mdr) [inline cloud/services/net/net.go:47]: It should be the cluster name, right ?

File: docs/develop.md [hunk 8]
```diff
@@ -103,6 +103,9 @@ cluster-api-provider-outscale-controller-manager-7d5c48d67t6d7f   2/2     Runnin
 
 # Develop
 
+## Install tools
+
+You can install also kind and kubebuildertool.
```
Comment (outscale-mdr) [inline docs/develop.md:108]: Link to these tools ?

**Review by outscale-mdr — CHANGES_REQUESTED** [review]
You should explain bug fixes in specific commits

File: Makefile [hunk 9]
```diff
@@ -111,16 +119,40 @@ cloud-init-secret:
...
+ccm-cni-ex: envsubst
+	$(ENVSUBST) < $(E2E_CONF_CLUSTER_CLASS_FILE_SOURCE) > $(E2E_CONF_CLUSTER_CLASS_FILE)
+
+.PHONY: cluster-class-ex
+cluster-class-ex: envsubst
+	$(ENVSUBST) < $(E2E_CLUSTER_CLASS_FILE_SOURCE) > $(E2E_CLUSTER_CLASS_FILE)
 
 .PHONY: test
-test: manifests generate fmt vet envtest ## Run tests.
+test: manifests generate fmt vet envtest ## Run tests.@
```
Comment (jerome-jutteau) [inline Makefile:166]: typo

File: api/v1beta1/oscmachine_validation.go [hunk 10]
```diff
@@ -134,9 +148,9 @@ func ValidateSubregionName(subregionName string) (string, error) {
 // ValidateDeviceName check that DeviceName  is a valid DeviceName
 func ValidateDeviceName(deviceName string) (string, error) {
 	last := deviceName[len(deviceName)-1:]
-	isValidateDeviceName := regexp.MustCompile(`^[a-z]$`).MatchString
+	isValidateDeviceName := regexp.MustCompile(`^[0-9a-z]$`).MatchString
 	switch {
-	case strings.HasPrefix(deviceName, "/dev/xvd") && len(deviceName) == 9 && isValidateDeviceName(last):
+	case (strings.Contains(deviceName, "/dev/xvd") || strings.Contains(deviceName, "/dev/sda")) && len(deviceName) == 9 && isValidateDeviceName(last):
```
Comment (jerome-jutteau) [inline api/v1beta1/oscmachine_validation.go:153]: if / else would be ok here.

You can combine all the logic in one regex in one human-readable regex I think

Also, you theoretically have "xvdaa" once "xvdz" is taken

File: api/v1beta1/oscmachine_webhook_test.go [hunk 11]
```diff
@@ -41,82 +41,71 @@ func TestOscMachine_ValidateCreate(t *testing.T) {
...
-							SubregionName: "eu-west-2a",
+					Vm: OscVm{
+						RootDisk: OscRootDisk{
+							RootDiskType: "gp3",
 						},
 					},
 				},
 			},
-			expValidateCreateErr: fmt.Errorf("OscMachine.infrastructure.cluster.x-k8s.io \"webhook-test\" is invalid: size: Invalid value: -30: Invalid size"),
+			expValidateCreateErr: fmt.Errorf("OscMachine.infrastructure.cluster.x-k8s.io \"webhook-test\" is invalid: diskType: Invalid value: \"gp3\": Invalid volumeType"),
```
Comment (jerome-jutteau) [inline api/v1beta1/oscmachine_webhook_test.go:108]: I would take an invalid type other than gp3.

File: api/v1beta1/types.go [hunk 12]
```diff
@@ -325,11 +333,12 @@ var (
 
 	DefaultClusterName string = "cluster-api"
 
+	DefaultVmName             string = "cluster-api-vm"
 	DefaultVmSubregionName    string = "eu-west-2a"
-	DefaultVmImageId          string = "ami-2fe74243"
+	DefaultVmImageId          string = "ami-e1a786f1"
```
Comment (jerome-jutteau) [inline api/v1beta1/types.go:338]: Is this OMI a duplicate from official one so we don't change it regularly ?

**Review by jerome-jutteau — CHANGES_REQUESTED** [review]

File: api/v1beta1/oscmachine_webhook_test.go (hunk 11 above)
Comment (outscale-vbr) [reply api/v1beta1/oscmachine_webhook_test.go:108]: ok done ssd1

File: api/v1beta1/types.go (hunk 12 above)
Comment (outscale-vbr) [reply api/v1beta1/types.go:338]: create from image builder and github action.
It will be change from omi name to image name.

File: api/v1beta1/oscmachine_validation.go (hunk 10 above)
Comment (outscale-vbr) [reply api/v1beta1/oscmachine_validation.go:153]: Ok done

File: .github/workflows/conformance.yaml (hunk 2 above)
Comment (outscale-vbr) [reply .github/workflows/conformance.yaml]: Ok change

File: api/v1beta1/oscmachine_webhook.go (hunk 3 above)
Comment (outscale-vbr) [reply api/v1beta1/oscmachine_webhook.go:67]: We can only update value which are in cluster-template.yaml.

File: api/v1beta1/types.go (hunk 4 above)
Comment (outscale-vbr) [reply api/v1beta1/types.go]: done

File: api/v1beta1/types.go (hunk 5 above)
Comment (outscale-vbr) [reply api/v1beta1/types.go]: done

File: cloud/services/net/net.go (hunk 7 above)
Comment (outscale-vbr) [reply cloud/services/net/net.go]: change to clusterName

File: docs/develop.md (hunk 8 above)
Comment (outscale-vbr) [reply docs/develop.md]: done

File: Makefile (hunk 9 above)
Comment (outscale-vbr) [reply Makefile]: done

File: api/v1beta1/types.go [hunk 13]
```diff
@@ -846,30 +840,20 @@ func (network *OscNetwork) SetSecurityGroupDefaultValue() {
...
 		}
-		
-		securityGroupRuleKwBgpKw := OscSecurityGroupRule{
-			Name: securityGroupRuleKwBgpKwName,
-			Flow: DefaultFlowKwBgp,
-			IpProtocol: DefaultIpProtocolKwBgp,
-			IpRange: DefaultRuleIpRangeKwBgp,
+
+		securityGroupRuleKwBgp := OscSecurityGroupRule{
+			Name:          securityGroupRuleKwBgpName,
```
Comment (outscale-hmi) [inline api/v1beta1/types.go:845]: white spaces

File: api/v1beta1/types.go [hunk 14]
```diff
@@ -908,29 +892,20 @@ func (network *OscNetwork) SetSecurityGroupDefaultValue() {
...
 		}
-		
-		securityGroupRuleKcpBgpKcp := OscSecurityGroupRule{
-			Name: securityGroupRuleKcpBgpKcpName,
-			Flow: DefaultFlowKcpBgp,
-			IpProtocol: DefaultIpProtocolKcpBgp,
-			IpRange: DefaultRuleIpRangeKcpBgp,
+
+		securityGroupRuleKcpBgp := OscSecurityGroupRule{
+			Name:          securityGroupRuleKcpBgpName,
```
Comment (outscale-hmi) [inline api/v1beta1/types.go:897]: white spaces

**Review by outscale-hmi — COMMENTED** [review]
some format issues

File: cloud/services/compute/vm.go [hunk 15]
```diff
@@ -162,3 +184,20 @@ func (s *Service) GetVmState(vmId string) (string, error) {
...
 	return *vmState, nil
 }
+
+// AddCcmTag add ccm tag
+func (s *Service) AddCcmTag(netName string, hostname string, vmId string) error {
+	resourceIds := []string{vmId}
+	oscApiClient := s.scope.GetApi()
+	oscAuthClient := s.scope.GetAuth()
+
+	err := tag.AddTag("OscK8SNodeName", hostname, resourceIds, oscApiClient, oscAuthClient)
```
[the remark this answers is not in the document]
Comment (outscale-hmi) [reply cloud/services/compute/vm.go:194]: in same tags you name for example OscK8sCluster and some palces you use K8S

File: test/e2e/capo.go [hunk 16]
```diff
@@ -117,10 +128,307 @@ func CapoClusterDeploymentSpec(ctx context.Context, inputGetter func() CapoClust
...
 	AfterEach(func() {
 		dumpSpecResourcesAndCleanup(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder, namespace, cancelWatches, clusterResources.Cluster, input.E2EConfig.GetIntervals, input.SkipCleanup)
 	})
 }
+
+func CapoClusterMachineDeploymentSpec(ctx context.Context, inputGetter func() CapoClusterMachineDeploymentSpecInput) {
+	var (
+		specName         = "capo"
+		input            CapoClusterMachineDeploymentSpecInput
+	//	namespace        *corev1.Namespace
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply test/e2e/capo.go]: Already remove

File: test/e2e/capo.go [hunk 17]
```diff
@@ -117,10 +128,307 @@ func CapoClusterDeploymentSpec(ctx context.Context, inputGetter func() CapoClust
...
+
+	BeforeEach(func() {
+		Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+		input = inputGetter()
+		Expect(input.E2EConfig).ToNot(BeNil(), "Invalid argument. input.E2EConfig can't be nil when calling %s spec", specName)
+		Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterConfigPath must be an existing file when calling %s spec", specName)
+		Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BoostrapClusterProxy can't be nil when calling %s spec", specName)
+		Expect(os.MkdirAll(input.ArtifactFolder, 0750)).To(Succeed(), "Invalid argument. input.ArtifactFolder can't be created for %s spec", specName)
+		Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersion))
+//		namespace, cancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder)
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply test/e2e/capo.go]: already remove

File: test/e2e/capo.go [hunk 18]
```diff
@@ -117,10 +128,307 @@ func CapoClusterDeploymentSpec(ctx context.Context, inputGetter func() CapoClust
...
+		utils.WaitForConfigMapsAvailable(ctx, utils.ConfigMapInput{
+			Getter:    k8sClient,
+			Name:      "harbor-chartmuseum",
+			Namespace: "harbor",
+		})
+
+		By("PASSED!")
+
+		
+	/*	clusterctl.ApplyClusterTemplateAndWait(ctx, clusterctl.ApplyClusterTemplateAndWaitInput{
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply test/e2e/capo.go]: already remove

File: test/e2e/capo.go [hunk 19]
```diff
@@ -117,10 +128,307 @@ func CapoClusterDeploymentSpec(ctx context.Context, inputGetter func() CapoClust
...
+		})
+		utils.WaitForDeleteSecretAvailable(ctx, utils.SecretInput{
+			Getter: workloadClient,
+			Name: "provisionner",
+			Namespace: "default",
+		})
+		By("PASSED!")
+	})
+	AfterEach(func() {
+//		dumpSpecResourcesAndCleanup(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder, namespace, cancelWatches, clusterResources.Cluster, input.E2EConfig.GetIntervals, input.SkipCleanup)
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply test/e2e/capo.go]: already remove

File: cloud/services/compute/vm.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply cloud/services/compute/vm.go:194]: https://github.com/outscale-dev/cloud-provider-osc/blob/280929766743910197228e7a8e4758de6923b663/deploy/README.md
it is OscK8SNodeName. I can not change it.

File: .github/workflows/kind.yaml (hunk 1 above)
Comment (outscale-vbr) [reply .github/workflows/kind.yaml]: I change it

File: cloud/services/compute/vm.go (hunk 6 above)
Comment (outscale-vbr) [reply cloud/services/compute/vm.go]: In futur pr yes
